clusterloader2/pkg/measurement/common/kube_state_metrics_measurement.go

```diff
@@ -0,0 +1,209 @@
+	if !m.isRunning {
+		klog.V(2).Infof("%s: measurement was not running", m)
+		return nil, nil
+	}
+	close(m.stopCh)
+	m.wg.Wait()
+	latestLatency, err := m.retrieveKSMLatencyMetrics(ctx, client)
+	if err != nil {
+		return nil, err
+	}
+	// We want to subtract the latest histogram from the first one we collect.
+	finalLatency := HistogramSub(latestLatency, m.initialLatency)
+	// Pretty Print the report.
+	result := &measurementutil.LatencyMetric{}
+	err = SetQuantileFromHistogram(result, finalLatency)
+	if err != nil {
+		return nil, err
+	}
+	content, err := util.PrettyPrintJSON(result)
+	if err != nil {
+		return nil, err
+	}
+	// Create Summary.
+	return []measurement.Summary{measurement.CreateSummary(ksmLatencyName, "json", content)}, nil
+}
+
+// startQuerying queries /metrics endpoint of kube-state-metrics kube_ metrics every interval
+// and stops when stop is called.
+func (m *ksmLatencyMeasurement) startQuerying(ctx context.Context, client clientset.Interface, interval time.Duration) {
+	m.isRunning = true
+	m.wg.Add(1)
+	go m.queryLoop(ctx, client, interval)
+}
+
+func (m *ksmLatencyMeasurement) queryLoop(ctx context.Context, client clientset.Interface, interval time.Duration) {
+	defer m.wg.Done()
+	for {
+		select {
+		case <-m.stopCh:
+			return
+		case <-ctx.Done():
+			return
+		case <-time.After(interval):
+			var output string
+			output, err := m.getMetricsFromService(ctx, client, m.metricsPort)
+			if err != nil {
+				klog.V(2).Infof("%s: measurement already running", m)
+			}
+			if output == "" {
+				klog.V(2).Infof("%s: measurement already running", m)
+				klog.V(2).Infof("/metrics endpoint of kube-state-metrics returned no data in namespace: %s from service: %s port: %d", m.namespace, m.serviceName, m.metricsPort)
+			}
+
+		}
+	}
+}
+
+func (m *ksmLatencyMeasurement) retrieveKSMLatencyMetrics(ctx context.Context, c clientset.Interface) (*measurementutil.Histogram, error) {
+	ksmHist := measurementutil.NewHistogram(nil)
+	output, err := m.getMetricsFromService(ctx, c, m.selfPort)
+	if err != nil {
+		return ksmHist, err
+	}
+	samples, err := measurementutil.ExtractMetricSamples(output)
+	if err != nil {
+		return ksmHist, err
+	}
+	for _, sample := range samples {
+		switch sample.Metric[model.MetricNameLabel] {
+		case ksmRequestDurationMetricName:
+			measurementutil.ConvertSampleToHistogram(sample, ksmHist)
+		}
+	}
+	return ksmHist, nil
+}
+
+func (m *ksmLatencyMeasurement) getMetricsFromService(ctx context.Context, client clientset.Interface, port int) (string, error) {
+	ctx, cancel := context.WithTimeout(ctx, 2*time.Minute)
+	defer cancel()
+	out, err := client.CoreV1().RESTClient().Get().
+		Resource("services").
+		SubResource("proxy").
+		Namespace(m.namespace).
+		Name(fmt.Sprintf("%v:%v", m.serviceName, port)).
+		Suffix("metrics").
+		Do(ctx).Raw()
+	return string(out), err
+}
+
+// Dispose cleans up after the measurement.
+func (m *ksmLatencyMeasurement) Dispose() {}
```

Comment:
Dispose should stop the measurement if it's still running.

So basically, you should probably:
- extract lines 114-119 to a separate stop() method
- call stop() there
- call stop here in Dispose method

Reply:
Thanks good to know! Will do. FYI I copied from other files, might be good to address it there as well.

Reply:
Many measurements don't start anything - they just do something on start and something on gather but they don't start any goroutines or anything like that (in which case nothing more is needed there).
There are certainly measurements that do that, e.g.
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go#L152